titan-compiler/checker.lua

```diff
@@ -476,19 +481,28 @@ checkvar = util.make_visitor({
     end,
 
     ["Ast.VarBracket"] = function(node, st, errors, context)
-        checkexp(node.exp1, st, errors, context and types.Array(context))
-        if node.exp1._type._tag ~= "Type.Array" then
+        checkexp(node.exp1, st, errors) -- FIXME check this: checkexp(node.exp1, st, errors, context and types.Array(context))
```

Comment:
This is saying that if `e1[e2]` is being used in a context that expects type `t` then check `e1` in a context that expects type `{ t }`. Right now only useful in things like `{ { foo = 2 } }[i]` where this context would be necessary to make sure that the type checker knows which record type the inner constructor refers to.

Now that brackets are overloaded for arrays and maps this breaks, as the context will be wrong with something like `{ ["rec"] = { foo = 2 } }[i]`. A hackish solution is to have a map type instead of array type as the context, and change the code that interprets the context in the type checking of the array constructor to know what to do with this map type.

Reply:
RIght now I can only think of contrived examples where passing a context to `e1` in `e1[e2]` is useful, so might as well remove this. It can come back later when/if there are real use cases. About the map vs. record difference, I also see both pros and cons, and would vote to keep it the way you did it for the time being.

Of course a separate type for "structural records" backed by Lua tables with string keys could use the same `{ foo = bar }` syntax as nominal records.